src/main/java/com/microsoft/sqlserver/jdbc/SQLServerSpatialDatatype.java

```diff
@@ -91,14 +97,118 @@
      * @param noZM
      *        flag to indicate if Z and M coordinates should be included
      */
-    protected abstract void serializeToWkb(boolean noZM);
+
+    protected void serializeToWkb(boolean noZM, SQLServerSpatialDatatype type) {
+        ByteBuffer buf = ByteBuffer.allocate(determineWkbCapacity());
+        createSerializationProperties();
+
+        buf.order(ByteOrder.LITTLE_ENDIAN);
+        buf.putInt(srid);
+        buf.put(version);
+        buf.put(serializationProperties);
+
+        if (!isSinglePoint && !isSingleLineSegment) {
+            buf.putInt(numberOfPoints);
+        }
+
+        if (type instanceof Geometry) {
+            for (int i = 0; i < numberOfPoints; i++) {
+                buf.putDouble(xValues[i]);
+                buf.putDouble(yValues[i]);
+            }
+        } else { // Geography
+            for (int i = 0; i < numberOfPoints; i++) {
+                buf.putDouble(yValues[i]);
+                buf.putDouble(xValues[i]);
+            }
+        }
+
+        if (!noZM) {
+            if (hasZvalues) {
+                for (int i = 0; i < numberOfPoints; i++) {
+                    buf.putDouble(zValues[i]);
+                }
+            }
+
+            if (hasMvalues) {
+                for (int i = 0; i < numberOfPoints; i++) {
+                    buf.putDouble(mValues[i]);
+                }
+            }
+        }
+
+        if (isSinglePoint || isSingleLineSegment) {
+            wkb = buf.array();
+            return;
+        }
+
+        buf.putInt(numberOfFigures);
+        for (int i = 0; i < numberOfFigures; i++) {
+            buf.put(figures[i].getFiguresAttribute());
+            buf.putInt(figures[i].getPointOffset());
+        }
+
+        buf.putInt(numberOfShapes);
+        for (int i = 0; i < numberOfShapes; i++) {
+            buf.putInt(shapes[i].getParentOffset());
+            buf.putInt(shapes[i].getFigureOffset());
+            buf.put(shapes[i].getOpenGISType());
+        }
+
+        if (version == 2 && null != segments) {
+            buf.putInt(numberOfSegments);
+            for (int i = 0; i < numberOfSegments; i++) {
+                buf.put(segments[i].getSegmentType());
+            }
+        }
+
+        if (noZM) {
+            wkbNoZM = buf.array();
+        } else {
+            wkb = buf.array();
+
+        }
+        return;
```

Comment:
Do we need this return statement here? Also a blank line in the else block above looks unnecessary.